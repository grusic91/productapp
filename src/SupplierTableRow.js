import React from 'react';

export class SupplierTableRow extends React.Component {
    render(){
        let s = this.props.supplier;
        return <tr>
            <td>{s.id}</td>
            <td>{s.name}</td>
            <td>{s.city}</td>
            <td>{s.products.join(", ")}</td>
            <td>
                <button 
                    className="btn btn-sm btn-warning m-1" 
                    onClick={() => this.props.editCallback(s)}
                >
                    Edit
                </button>
                <button 
                    className="btn btn-sm btn-danger m-1" 
                    onClick={() => this.props.deleteCallback(s)}
                >
                    Delete
                </button>
            </td>
        </tr>
    }
}