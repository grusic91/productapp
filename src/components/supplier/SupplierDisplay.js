import React from 'react';
import PropTypes from 'prop-types';
import { SupplierEditor } from './SupplierEditor';
import { SupplierTable } from './SupplierTable';

export class SupplierDisplay extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            showEditor: false,
            selected: null
        }
    }

    startEditing = supplier => {
        this.setState({
            showEditor: true,
            selected: supplier
        })
    }

    createSupplier = () => {
        this.setState({
            showEditor: true,
            selected: {}
        })
    }

    cancelEditing = () => {
        this.setState({
            showEditor: false,
            selected: null
        })
    }

    saveSupplier = supplier => {
        this.props.saveCallback(supplier);
        this.setState({
            showEditor: false,
            selected: null
        });
    }

    render() {
        const { showEditor, selected } = this.state;

        return (
            <div className="m-2">
            { showEditor ? 
                <div>
                    <SupplierEditor
                        key={selected.id || -1 }
                        supplier={selected}
                        saveCallback={this.saveSupplier}
                        cancelEditing={this.cancelEditing}
                    />
                </div>
                    :
                <React.Fragment>
                    <SupplierTable
                        suppliers={this.props.suppliers}
                        editCallback={this.startEditing}
                        deleteCallback={this.props.deleteCallback}
                    />
                    <div className="text-center">
                        <button className="btn btn-primary m-1" onClick={this.createSupplier}>
                            Create Supplier
                        </button>
                    </div>
                </React.Fragment>
            }
            </div>
        )
    }
}


SupplierDisplay.propTypes = {
    name: PropTypes.string.isRequired,
    suppliers: PropTypes.array.isRequired,
    saveCallback: PropTypes.func.isRequired,
    deleteCallback: PropTypes.func.isRequired,
}