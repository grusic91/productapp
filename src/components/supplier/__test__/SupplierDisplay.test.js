import React from 'react';
import { shallow, mount } from 'enzyme';
import { SupplierDisplay } from '../SupplierDisplay';

import { SupplierTable } from '../SupplierTable';
import { SupplierEditor } from '../SupplierEditor';
import { SupplierTableRow } from '../SupplierTableRow';

describe('<SupplierDisplay />', () => {
    let wrapper;
    beforeAll(() => wrapper = shallow(<SupplierDisplay 
        name={""}
        suppliers={[]}
        saveCallback={jest.fn()}
        deleteCallback={jest.fn()}
    />));

    it("renders properly", () => {
        expect(wrapper).toMatchSnapshot();
    })

    it("should render two divs elements", () => {
        expect(wrapper.find('div')).toHaveLength(2);
    });

    it("conditional Render SupplierTable", () => {
        wrapper.setState({showEditor: false});
        expect(wrapper.containsMatchingElement(<SupplierTable />)).toEqual(true);
    });

    it("conditional Rendering Create Supplier <button> element", () => {
        wrapper.setState({ showEditor: false});
        expect(wrapper.find('button').text()).toEqual("Create Supplier");
    });

    it("conditional Rendering SuplierEditor", () => {
        wrapper.setState({ 
            showEditor: true, 
            selected: {
                id: "",
                name: "",
                city: "",
                products: []
            }
        });
        expect(wrapper.containsMatchingElement(<SupplierEditor />)).toEqual(true);
    });
});

describe('<SupplierDisplay /> functionality', () => {
    let wrapper;

    const saveData = (collection, item) => {
        if (item.id === "") {
            item.id = this.idCounter++;
            this.setState(state => state[collection] = state[collection].concat(item))
        } else {
            this.setState(state => state[collection] = state[collection].map(
                stored => stored.id === item.id ? item: stored
            ))
        }
    }

    const deleteData = (collection, item) => {
        this.setState(state => state[collection] = state[collection].filter(
            stored => stored.id !== item.id
        ));
    }

    beforeAll(() => wrapper = mount(<SupplierDisplay 
        name={"Suppliers"}
        suppliers={[
            {
                id: 1,
                name: "Surf Dudes",
                city: "San Jose",
                products: [1, 2]
            },
            {
                id: 2,
                name: "Field Supplies",
                city: "New York",
                products: [3]
            }
        ]}
        saveCallback={saveData}
        deleteCallback={deleteData}
    />));
    
    it("click 'Edit' button in supplier table, ProductEditor component should be rendered", () => {
        const supplierTable = wrapper.find(SupplierTable);
        const editButton = supplierTable.find(SupplierTableRow).at(0).find('button').at(0);
        editButton.simulate('click');

        expect(wrapper.containsMatchingElement(SupplierEditor)).toEqual(true);
    });

});