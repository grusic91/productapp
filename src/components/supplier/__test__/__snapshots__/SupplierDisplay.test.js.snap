// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SupplierDisplay /> renders properly 1`] = `ShallowWrapper {}`;
